src/transforms/aggregate.rs

```diff
@@ -0,0 +1,440 @@
+        let metric = event.as_metric();
+        let series = metric.series();
+        let data = metric.data();
+        let map = &mut self.map;
+
+        match data.kind {
+            metric::MetricKind::Incremental => {
+                match map.get_mut(&series) {
+                    // We already have something, add to it, will update timestamp as well.
+                    Some(existing) => existing.update(data),
+                    None => {
+                        // New so store
+                        map.insert(series.clone(), data.clone());
+                        true
+                    }
+                };
+            },
+            metric::MetricKind::Absolute => {
+                // Always replace/store
+                map.insert(series.clone(), data.clone());
+            }
+        };
+
+        emit!(AggregateEventRecorded);
+    }
+
+    fn flush_into(&mut self, output: &mut Vec<Event>) -> u64 {
+        let mut count = 0_u64;
+        let map = &mut self.map;
+
+        if map.len() > 0 {
+            // TODO: not clear how this should work with aggregation so just stuffing a default one
+            // in for now.
+            let metadata = EventMetadata::default();
+
+            for (series, metric) in map.drain() {
+                let metric = metric::Metric::from_parts(series, metric, metadata.clone());
+                output.push(Event::Metric(metric));
+                count += 1;
+            }
+        }
+
+        emit!(AggregateFlushed);
+        return count;
+    }
+}
+
+impl TaskTransform for Aggregate {
+    fn transform(
+        self: Box<Self>,
+        mut input_rx: Pin<Box<dyn Stream<Item = Event> + Send>>,
+    ) -> Pin<Box<dyn Stream<Item = Event> + Send>>
+    where
+        Self: 'static,
+    {
+        let mut me = self;
+
+        let interval = me.interval;
+
+        let mut flush_stream = tokio::time::interval(interval);
+
+        Box::pin(
+            stream! {
+                loop {
+                    let mut output = Vec::new();
+                    let done = tokio::select! {
+                        _ = flush_stream.tick() => {
+                            me.flush_into(&mut output);
+                            false
+                        }
+                        maybe_event = input_rx.next() => {
+                            match maybe_event {
+                                None => {
+                                    me.flush_into(&mut output);
+                                    true
+                                }
+                                Some(event) => {
+                                    me.record(event);
+                                    false
+                                }
+                            }
+                        }
+                    };
+                    yield stream::iter(output.into_iter());
+                    if done { break }
+                }
+            }
+            .flatten(),
```

Comment:
Yeah. I actually started down the path of rewirting this at one point b/c it was hard to follow, but then rememembered it was a direct copy-n-past from reducer so let it alone so that the implementations would match, e.g. a bug or change in one would directly apply to the other. If you don't care if they're dissimilar i can make this change. https://github.com/timberio/vector/blob/41409e7b6d1049491e1c591cbea6ba3b49222ed6/src/transforms/reduce/mod.rs#L263-L290

Reply:
Yeah, the implementation in reduce is just, IMO, suboptimal.  We have a chance here to improve on that original decision, so that's the driving force here.